themes/bootstrap3/templates/RecordDriver/SolrOverdrive/core.phtml

```diff
@@ -0,0 +1,141 @@
+<div class="media" vocab="http://schema.org/" resource="#record" typeof="<?php echo $this->driver->getSchemaOrgFormats()?> Product">
```

Comment:
Just food for thought: I wonder if there's a way to implement this without having to duplicate all of the code from DefaultRecord/core.phtml. Again, I haven't done a close comparison, so maybe it's unavoidable to have to do it this way... but if we could achieve the same results by either wrapping extra markup around a render of the default template, or if we could create an optional placeholder variable in DefaultRecord/core that we could pass in from here, that might simplify future code maintenance. This can stand as-is if that's a crazy or impractical idea, though!

Reply:
(And ditto for all of the other custom record driver templates as well, where applicable).

Reply:
I don't have strong feelings about this.  I did try to keep the templates fairly similar to the default for consistency.  But it could be that users would want to to modify to make them different.  I'm willing to try something else if you want to give me more direction on it.  Otherwise, I'm going to continue down this path for now.

Reply:
We can stick with this for now and revisit later if it proves to be a maintenance burden.

Reply:
OK, I'm changing my tune on this.  I already have changes coming through on the core template that I didn't realize were there.  Most of the markup I have so far is in one discrete chunk.  Based on that it would seem that your  "optional placeholder" idea would make more sense.  I feel like in older versions of VF you could just include another template.  Is that what you are thinking?  Or maybe a view helper?

Reply:
Exactly. You can $this->render('path/to/template') to include another template... so you could, hypothetically, do something like:

<pre>
$this->extraContent = "my custom html";
echo $this->render('RecordDriver/DefaultRecord/core.phtml');
</pre>

If the default core.phtml were modified to do something when $this->extraContent is set.

(Contrived example, of course -- but I think you see my idea).

Reply:
Now I didn't check the modifications required to the templates, but here's one idea anyway: One option would also be to include predefined extension points in the default templates like this:

    ...
    <?php if ($this->resolver('RecordDriver/DefaultRecord/record-pre-toolbar.phtml')): ?>
      <?=$this->render('RecordDriver/DefaultRecord/record-pre-toolbar.phtml') ?>
    <?php endif; ?>

    <?=$this->record($this->driver)->getToolbar()?>

    <?php if ($this->resolver('RecordDriver/DefaultRecord/record-post-toolbar.phtml')): ?>
      <?=$this->render('RecordDriver/DefaultRecord/record-post-toolbar.phtml') ?>
    <?php endif; ?>
    ...

The three lines above could be put in a helper so that they'd become a single line of template code (like `$this->renderIfExists('...'); ?>`. Now I have no idea if this has performance implications if used a lot, but I suppose it's lighter than always including a set of templates that by default would be empty.

Reply:
From a design clarity perspective, I definitely prefer @EreMaijala's approach. However, from a performance perspective, my suggestion of injecting content through a variable is probably preferable. In order for the resolver to decide whether or not a template is available, it has to search a significant number of different possible filenames (number of themes in the inheritance chain multipled by number of record drivers in the class hierarchy), which, when combined with multiple extension points and multiple records being displayed in a result set, adds up to a LOT of file system reads. I haven't done benchmarking to see how much this matters in practice, but my gut feeling is that it could be problematic at least under some conditions.

Regardless of whether you take the "pull" or "push" approach to the problem, the most important thing is picking good extension points where content could be inserted. Then it becomes a matter of either overriding the correct templates (if using Ere's way) or setting the correct variables (if using my way). If we design the extension points well, then the nature of the implementation could be changed in the future with relatively little fuss.

Reply:
What about a way to figure out the record type ( Essentially the string that is used to resolve the driver.?)

<? if ($this->record($this->driver)->getType() == "overdrive"): 
      <?= $this->render('RecordDriverOverdrive/Solr/core.phtml'); ?>
<? endif;?>

Reply:
My goal is to avoid any solution that adds hard-coded logic about specific drivers into code intended to be used generically. That's the sort of thing that's okay for a while but can grow into an unmanageable mess over time. (One special case if statement seems fine... but eventually it can end up being twenty). That's why I prefer a solution that introduces hooks of some sort and allows them to be used by the more specific code.

Reply:
Note that you always have the option of using `$this->record($this->driver)->renderTemplate($templateName)`, and the code will look for `$templateName` in the template folder specific to your record driver, then look up the inheritance chain until it finds a match. So you can create a hook by building an empty template in the DefaultRecord templates and overriding it in Overdrive. As I mentioned to Ere, I have concerns that doing this kind of thing too much will impact performance... but if that's the easiest way to approach it for now, feel free to try it in this pull request. It will help figure out the logical separations we need, and we can adjust the implementation as needed before we merge to master.

Reply:
That's strange; I thought that render automatically inherited the current context, so I don't see why that isn't working unless I'm forgetting some fundamental detail. Out of curiosity, does it work differently if you try:

`echo $this->context()->renderInContext('RecordDriver/DefaultRecord/core.phtml', ['OverDriveMarkup => '<p>my markup</p>']);`

?

I don't think you should have to use the context helper for this, but I'd be interested to see if it makes a difference. If nothing else, that might offer a clue.

Reply:
Well I actually had two values to pass through and I was trying to save them in an array.   for some reason, it didn't like that.  So when I changed it to just two different variables it worked.

Reply:
Hmm, I'm not sure exactly why an array would be a problem... but if two variables works, that's probably fine too. Let me know if you need to dig deeper into this!